File: whistleblower/contracts-java/src/main/java/net/corda/examples/whistleblower/states/BlowWhistleState.java
package net.corda.examples.whistleblower.states;

import com.google.common.collect.ImmutableList;
import net.corda.core.contracts.BelongsToContract;
import net.corda.core.contracts.LinearState;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.AbstractParty;
import net.corda.core.identity.AnonymousParty;
import net.corda.core.identity.Party;
import net.corda.examples.whistleblower.contracts.BlowWhistleContract;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * A state representing a whistle-blowing case.
 *
 * The identity of both the whistle-blower and the investigator is kept confidential through the
 * use of [AnonymousParty].
 *
 * @property badCompany the company the whistle is being blown on.
 * @property whistleBlower the [AnonymousParty] blowing the whistle.
 * @property investigator the [AnonymousParty] handling the investigation.
 */
@BelongsToContract(BlowWhistleContract.class)
public class BlowWhistleState implements LinearState {
    private final Party badCompany;
    private final AnonymousParty whistleBlower;
    private final AnonymousParty investigator;
    private final UniqueIdentifier linearId = new UniqueIdentifier();

    public BlowWhistleState(Party badCompany, AnonymousParty whistleBlower, AnonymousParty investigator) {
        this.badCompany = badCompany;
        this.whistleBlower = whistleBlower;
        this.investigator = investigator;
    }

    public Party getBadCompany() {
        return badCompany;
    }

    public AnonymousParty getWhistleBlower() {
        return whistleBlower;
    }

    public AnonymousParty getInvestigator() {
        return investigator;
    }

    @NotNull
    @Override
    public UniqueIdentifier getLinearId() {
        return linearId;
    }

    @NotNull
    @Override
    public List<AbstractParty> getParticipants() {
        return ImmutableList.of(whistleBlower, investigator);
    }
}

File: whistleblower/kotlin-version/src/main/kotlin/com/whistleblower/state/BlowWhistleState.kt
package com.whistleblower.state

import com.whistleblower.contract.BlowWhistleContract
import net.corda.core.contracts.BelongsToContract
import net.corda.core.contracts.LinearState
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.identity.AnonymousParty
import net.corda.core.identity.Party

/**
 * A state representing a whistle-blowing case.
 *
 * The identity of both the whistle-blower and the investigator is kept confidential through the
 * use of [AnonymousParty].
 *
 * @property badCompany the company the whistle is being blown on.
 * @property whistleBlower the [AnonymousParty] blowing the whistle.
 * @property investigator the [AnonymousParty] handling the investigation.
 */
@BelongsToContract(BlowWhistleContract::class)
data class BlowWhistleState(
        val badCompany: Party,
        val whistleBlower: AnonymousParty,
        val investigator: AnonymousParty,
        override val linearId: UniqueIdentifier = UniqueIdentifier()) : LinearState {

    // Both the whistle-blower and the current investigator are kept in the loop.
    override val participants = listOf(whistleBlower, investigator)
}